caching_redirect.go

```diff
@@ -0,0 +1,43 @@
+package main
+
+import (
+	"github.com/coocood/freecache"
+	"log"
+)
+
+const cacheExpiration = 86400 * 7 // 7 days
+
+type CachingRedirectClient struct {
+	redirectClient *RedirectClient
+	cache          *freecache.Cache
+}
+
+func NewCachingRedirectClient(apiKey string, baseId string) (*CachingRedirectClient, error) {
+	client, err := NewRedirectClient(apiKey, baseId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CachingRedirectClient{redirectClient: client, cache: freecache.NewCache(1024)}, nil
+}
+
+func (client *CachingRedirectClient) Get(key string) (*Redirect, error) {
+	val, err := client.cache.Get([]byte(key))
+	if err == nil {
+		return &Redirect{key, string(val)}, nil
+	}
+
+	redirect, err := client.redirectClient.Get(key)
+	if err != nil {
+		return redirect, err
+	}
+
+	log.Println("Caching", key, "->", redirect.URL)
```

Comment:
Don't use the global logger. Accept a `*log.Logger` as an argument in the constructor.

Also, with this many arguments, it may be preferable to build a struct to construct a redirecting client:

```
type CachingRedirectClientConfig struct {
  RedirectClient *redirect.Client // this can be an interface
  Cache *freecache.Cache  // this can be an interface
  Logger *log.Logger
```

Reply:
Also,

```
log.Println("Caching key:", redirect.URL)
```